### ISSUE TITLE
Migration create table generator does not have timestamps

### ISSUE THREAD

**Issue by msducheminjr** [rails/rails#28706]
### Steps to reproduce

Discovered while trying to write a custom generator that invokes the migration generator, but can be seen in the simplest of examples.

In any Rails application, use the Rails migration generator to create a table.
```bash
$ rails generate migration CreateArticles title:string:index body:text
```

### Expected behavior
The expected behavior is that this create table migration will create the table with timestamps unless you pass --no-timestamps or --skip-timestamps to the generator. The expected migration should match what you get when you use the model generator.  

```bash
$ rails generate model Article title:string:index body:text
```

This produces
```ruby
# db/migrate/[migration_number]_create_articles.rb
class CreateArticles < ActiveRecord::Migration[5.0]
  def change
    create_table :articles do |t|
      t.string :title
      t.text :body

      t.timestamps
    end
    add_index :articles, :title  
end
````

One would expect that the migration generator would have the same behavior, especially since they share the same [migration template](https://github.com/rails/rails/blob/master/activerecord/lib/rails/generators/active_record/migration/templates/create_table_migration.rb):

```ruby
# activerecord/lib/rails/generators/active_record/migration/templates/create_table_migration.rb:13-15
<% if options[:timestamps] %>
      t.timestamps
<% end -%>
```

The [model generator](https://github.com/rails/rails/blob/master/activerecord/lib/rails/generators/active_record/model/model_generator.rb) implements a class option for timestamps:

```ruby
 # activerecord/lib/rails/generators/active_record/model/model_generator.rb:11
 class_option :timestamps, type: :boolean
```

And has [test coverage in the railties generator tests](https://github.com/rails/rails/blob/master/railties/test/generators/model_generator_test.rb):

```ruby
# railties/test/generators/model_generator_test.rb:254-267
  def test_migration_with_timestamps
    run_generator
    assert_migration "db/migrate/create_accounts.rb", /t.timestamps/
  end

  def test_migration_timestamps_are_skipped
    run_generator ["account", "--no-timestamps"]

    assert_migration "db/migrate/create_accounts.rb" do |m|
      assert_method :change, m do |up|
        assert_no_match(/t.timestamps/, up)
      end
    end
  end
```

### Actual behavior
One would expect the following rails generate commands to produce equivalent migrations:

```bash
$ rails generate model Article title:string:index body:text
$ rails generate migration CreateArticles title:string:index body:text
```

But the second command will produce the following migration file:

```ruby
# db/migrate/[migration_number]_create_articles.rb
# No Timestamps produced
class CreateArticles < ActiveRecord::Migration[5.0]
  def change
    create_table :articles do |t|
      t.string :title
      t.text :body
    end
    add_index :articles, :title  
end
````

Likewise, these commands should produce equivalent migrations:

```bash
$ rails generate model Article title:string:index body:text --no-timestamps
$ rails generate migration CreateArticles title:string:index body:text --no-timestamps
```

If you add similar timestamp tests to the [migration generator test](https://github.com/rails/rails/blob/master/railties/test/generators/migration_generator_test.rb) railties/test/generators/migration_generator_test.rb file, the one expecting timestamps fails (and the one expecting them to be skipped passes by accident).

```ruby
# proposed additions to railties/test/generators/migration_generator_test.rb
def test_create_table_migration_with_timestamps
  run_generator ["create_books", "title:string", "content:text"]
  assert_migration "db/migrate/create_books.rb", /t.timestamps/
end

def test_create_table_timestamps_are_skipped
  run_generator ["create_books", "title:string", "content:text", "--no-timestamps"]

  assert_migration "db/migrate/create_books.rb" do |m|
    assert_method :change, m do |change|
      assert_no_match(/t.timestamps/, change)
    end
  end
end
```

### Solution

Adding a class option to activerecord/lib/rails/generators/active_record/migration/migration_generator.rb for timestamps. Solves the problem and the tests added to the railties migration tests now pass. I ran railties/test/generators/* and the whole ActiveRecord test suite on the [rails-dev-box](https://github.com/rails/rails-dev-box) and they pass.

```ruby
 # activerecord/lib/rails/generators/active_record/migration/migration_generator.rb:11
 class_option :timestamps, type: :boolean
```

I ran the full rails test suite because even though it's one line of code, changing a rails generator could have unexpected consequences and it's my first contribution.

### System configuration
**Rails version**: 5.0.1, master
I haven't tested it in every Rails version, but it looks like it would be an issue for any version greater than 3.0 from cycling through the branches for the activerecord/lib/rails/generators/active_record/model/model_generator.rb and activerecord/lib/rails/generators/active_record/migration/migration_generator.rb files.

**Ruby version**: 2.4.0

**Comment by rafaelfranca** [comment]
@msducheminjr could you open a PR?

**Comment by msducheminjr** [comment]
Working on it now.

**Comment by msducheminjr** [comment]
@rafaelfranca Issued #28707

**Comment by matthewd** [comment]
I disagree that this is expected behaviour: a join table, for example, would often not contain timestamps.

Directly building a create table migration is a lower level operation, so to me, it follows that it should make less assumptions about how the result should look: "I want a model to store \<this data>" is more open to interpretation and conventions; reinterpreting "I want a table with \<these columns>" seems more surprising.

I can see the counterargument if you view the migration generator as "do the migration-generating part of the model generator", but that's not really what it is, IMO.

**Comment by msducheminjr** [comment]
I agree with you that it is only the expected behavior for a create table. Right now as it is, the migration generator recognizes whether the migration generation is a create_table and uses the active_record/migration/templates/create_table_migration.rb template. It's the same template that the model generator uses and expects a class option for timestamps. Right now it is impossible to get timestamps using rails g migration and you need to go in and manually edit the migration. If you had the opportunity on the command line to pass a class_option boolean like in the model generator, you would have aligned behavior and expectations.

Even if we defaulted it to false and the class option was available, it would improve it for somebody who wanted that behavior from the command line.

**Comment by msducheminjr** [comment]
If we went this way, it wouldn't break anything backwards in terms of expectations and it would give users the ability to get timestamps into a migration if they want to.

```ruby
 # activerecord/lib/rails/generators/active_record/migration/migration_generator.rb
 class_option :timestamps, type: :boolean, default: false
```

Tests:
```ruby
# proposed additions to railties/test/generators/migration_generator_test.rb
def test_create_table_migration_with_no_timestamps_argument
  run_generator ["create_books", "title:string", "content:text"]
  assert_migration "db/migrate/create_books.rb" do |m|
    assert_method :change, m do |change|
      assert_no_match(/t.timestamps/, change)
    end
  end  
end

def test_create_table_migration_with_timestamps_argument
  run_generator ["create_books", "title:string", "content:text", "--timestamps"]
  assert_migration "db/migrate/create_books.rb", /t.timestamps/
end

def test_create_table_timestamps_are_explicitly_skipped
  run_generator ["create_books", "title:string", "content:text", "--no-timestamps"]

  assert_migration "db/migrate/create_books.rb" do |m|
    assert_method :change, m do |change|
      assert_no_match(/t.timestamps/, change)
    end
  end
end
```

**Comment by msducheminjr** [comment]
By the way, thank you for the extremely quick acknowledgment and feedback, @rafaelfranca! Within 30 minutes of submitting the issue there was thoughtful input on it. Really impressed.

**Comment by msducheminjr** [comment]
@matthewd , I did more research on this, and adding the line:

```ruby
class_option :timestamps, type: :boolean
```

to the [ActiveRecord migration generator](https://github.com/rails/rails/blob/2b4583f2a2ef94bc8c046e36e2b62ec642bbfb41/activerecord/lib/rails/generators/active_record/migration/migration_generator.rb) will have no effect on the join_table use case because /join_table/ comes before /^create_(.+)/ in the [case statement in set_local_assigns!](https://github.com/rails/rails/blob/2b4583f2a2ef94bc8c046e36e2b62ec642bbfb41/activerecord/lib/rails/generators/active_record/migration/migration_generator.rb#L23-43).

The only situation that gets the create_table_migration.rb template (which expects a class option for timestamps) is a create at the beginning of the command line argument without the join_table anywhere in the name.

If you add this 

```ruby
 # activerecord/lib/rails/generators/active_record/migration/migration_generator.rb:11
 class_option :timestamps, type: :boolean
```

All of the examples below pass:
```ruby
  def test_create_join_table_migration
    migration = "add_media_join_table"
    run_generator [migration, "artist_id", "musics:uniq"]

    assert_migration "db/migrate/#{migration}.rb" do |content|
      assert_method :change, content do |change|
        assert_match(/create_join_table :artists, :musics/, change)
        assert_match(/# t.index \[:artist_id, :music_id\]/, change)
        assert_match(/  t.index \[:music_id, :artist_id\], unique: true/, change)
        # line below added by me
        assert_no_match(/t.timestamps/, change)
      end
    end
  end

  # completely new example
  def test_no_timestamps_if_create_join_table_in_migration_name
    migration = "create_join_table_media_musics"
    run_generator [migration, "artist_id", "musics:uniq"]

    assert_migration "db/migrate/#{migration}.rb" do |content|
      assert_method :change, content do |change|
        assert_match(/create_join_table :artists, :musics/, change)
        assert_match(/# t.index \[:artist_id, :music_id\]/, change)
        assert_match(/  t.index \[:music_id, :artist_id\], unique: true/, change)
        assert_no_match(/t.timestamps/, change)
      end
    end

  end
```

For the create_table use case \(and __only__ the create_table use case\) you would have timestamps by default with the option to skip them at the command line.

**Comment by msducheminjr** [comment]
Fixed by #28707.
